src/app/components/chat/window/send/form/AppChatWindowSendForm.vue

```diff
@@ -0,0 +1,503 @@
+		return;
+	}
+
+	await submitMessage();
+
+	disableTypingTimeout();
+
+	// Refocus editor after submitting message with enter.
+	focusToken.focus();
+
+	applyNextMessageTimeout();
+}
+
+function applyNextMessageTimeout() {
+	// private
+	if (!room.value.isFiresideRoom) {
+		return;
+	}
+
+	const { currentUser } = chat.value;
+	// For fireside rooms, timeout the user from sending another message for 1.5s.
+	// Do not do this for the owner/mods.
+	if (currentUser?.id === room.value.owner_id) {
+		return;
+	}
+
+	if (currentUser) {
+		const userRole = tryGetRoomRole(chat.value, room.value, currentUser);
+		if (userRole === 'owner' || userRole === 'moderator') {
+			return;
+		}
+	}
+
+	nextMessageTimeout.value = setTimeout(() => {
+		if (nextMessageTimeout.value) {
+			clearTimeout(nextMessageTimeout.value);
+			nextMessageTimeout.value = null;
+		}
+	}, 1500);
+}
+
+function onChange(_value: string) {
+	if (!typing.value) {
+		typing.value = true;
+		startTyping(chat.value, room.value);
+	} else if (typingTimeout) {
+		clearTimeout(typingTimeout);
+	}
+	typingTimeout = setTimeout(disableTypingTimeout, TYPING_TIMEOUT_INTERVAL);
+}
+
+function onFocusEditor() {
+	isEditorFocused.value = true;
+	emit('focusChange', true);
+}
+
+function onBlurEditor() {
+	isEditorFocused.value = false;
+	emit('focusChange', false);
+}
+
+function onTabKeyPressed() {
+	if (!isEditorFocused.value) {
+		focusToken.focus();
+	}
+}
+
+function onUpKeyPressed(event: KeyboardEvent) {
+	if (isEditing.value || hasContent.value) {
+		return;
+	}
+	const { messages, currentUser } = chat.value;
+
+	// Find the last message sent by the current user.
+	const userMessages = messages[room.value.id].filter(msg => msg.user.id === currentUser?.id);
+	const lastMessage = userMessages[userMessages.length - 1];
+
+	if (lastMessage) {
+		// Prevent the "up" key press. This is to stop it from acting as a "go to beginning of line".
+		// The content editor is focused immediately after this, and we want the editor to focus the end
+		// of the content. This prevents it jump to the beginning of the line.
+		event.preventDefault();
+
+		setMessageEditing(chat.value, lastMessage);
+	}
+}
+
+async function cancelEditing() {
+	emit('cancel');
+	setMessageEditing(chat.value, null);
+	clearMsg();
+
+	// Wait in case the editor loses focus
+	await nextTick();
+	// Regain focus on the editor
+	focusToken.focus();
+}
+
+async function clearMsg() {
+	form.formModel.content = '';
+	form.formModel.id = undefined;
+
+	// Wait for errors, then clear them.
+	await nextTick();
+	form.clearErrors();
+}
+
+function disableTypingTimeout() {
+	typing.value = false;
+	stopTyping(chat.value, room.value);
+}
+</script>
+
+<template>
+	<AppForm :controller="form">
+		<AppShortkey shortkey="tab" @press="onTabKeyPressed" />
+
+		<transition name="fade">
+			<div
+				v-if="!!typingText || isEditing"
+				class="-top-indicators"
+				:class="{
+					'-light-mode': !themeStore.isDark,
```

Comment:
i don't think it's gonna unwrap this. it's a ref on the object and templates only shallow unwrap. you should destructure this out